### DISCUSSION TITLE
스프링 로깅 전략

### DISCUSSION THREAD

**Discussion by jminkkk** [woowacourse-teams/2024-code-zap#200] [기록해봐요]
# 스프링 로깅 전략 결정
## 어떻게 로그를 관리할 것인지

**Comment by jminkkk** [comment]
## 어떤 로그를 남길 것인가
> 모든 로그에는 시간을 출력한다.

### request 정보
  - [x] request 별 쓰레드 식별값, 엔드포인트, http method
### 에러 정보
  - [x] 500대: 스택 트레이스
### response 정보
  - [x] 모든 응답 status, 응답 시간, 응답 바디로 통일
  - [x] ~~200대, 500대: 응답 상태코드~~
  - [ ] ~~400대: 예외 메시지~~
### 메서드 실행 정보
  - [x] 실행된 메서드 이름 및 실행 시간
### 트랜잭션 정보 로깅

**Comment by jminkkk** [comment]
# 어떤 로깅 프레임워크를 사용할 것인가?
## 고려한 스택과 그 이유
### Log4j
고려하지 않음 
+ 이유
  + 지원 중단
   + 보안 취약점에 대해 이슈가 있었음
   + 성능적으로도 가장 좋지 않음

### Logback
+ spring에서 표준인 로깅 프레임워크
+ 별도의 추가 의존성 없이도 사용할 수 있음
+ 자동 리로딩 등 여러 기능을 제공

###  log4j2
+ 멀티 쓰레드 환경에서 성능상으로 가장 유리함
+ 사용 시 spring에서 기본적으로 설정되어 있는 logback 의존성을 제거하고 사용해야 함

## 결론
### Logback을 사용한다.
> lms 요구사항: 서비스의 규모를 고려하여 러닝 커브가 큰 기술을 이용하여 구성하는 것 보다 간단하게나마 확인할 수 있는 정도로만 구성하기

현재 우리 서비스는 규모가 크지 않기 때문에 Logback을 사용해도 성능상 충분할 것이라 생각

또한, 별도의 의존성을 고려하지 않아도 이미 추가된 `spring-boot-starter-web`에 포함되어 있음.
<img width="1330" alt="image" src="https://github.com/user-attachments/assets/5e5f3aa6-7829-4e09-a7d3-67432a069a5c">

[스프링 공식 문서 참고](https://docs.spring.io/spring-boot/how-to/logging.html)
> Spring Boot has a LoggingSystem abstraction that attempts to configure logging based on the content of the classpath. If Logback is available, it is the first choice.
>
> Spring Boot에는 클래스 경로의 내용을 기반으로 로깅을 구성하려고 시도하는 LoggingSystem 추상화가 있습니다. Logback을 사용할 수 있는 경우, 이것이 **첫 번째 선택입니다**.

**Comment by jminkkk** [comment]
# 환경별 로깅 레벨 정하기

효과적인 로그 관리 및 디버깅을 위해 각 환경에 맞는 로깅 전략이 필요합니다.

## 운영 환경별 로깅 레벨 수준이 달라야 하는 이유

1. 예를 들어, dev 환경 또는 production 환경에서 모든 수준의 로그가 콘솔에 출력된다면 어떤 일이 발생할까요?
애플리케이션이 종료되지 않는 한 무분별하게 로그들이 출력될 것입니다. 무수히 많은 정보들로 인해 정작 확인해야 할 정보를 찾기 힘들 수 있습니다.
또 배포된 인스턴스의 어딘가에 저장될 이 로그들이 너무 많다면 관리해야 할 파일이 늘어나고 그 용량이 부담될 수 있습니다.

2. 반면에, 또 local 단계에서 로그를 막게 된다면 어떨까요?
내가 작성한 쿼리메서드가 어떤 쿼리로 나가는지 확인하지 못한다면? 로컬에서 디버깅을 위해 에러 상황을 재현하려는 정보를 얻을 수 없다면? 등등...
개발 시에 애플리케이션 동작을 빠르게 모니터링하고 문제를 빠르게 파악할 수 있어야 하는데, 내가 원하는 로그를 확인할 수 없다면 불편할 거에요.

따라서 우리는 **각 환경별 특성에 맞게 로그 레벨을 결정**해야 합니다.

## 환경별 로그 레벨

레벨을 결정하기 전, 로그백을 사용하는 우리는 5단계의 로그 레벨을 활용할 수 있습니다.
ERROR, WARN, INFO, DEBUG, TRACE 인데요. 
~~(다른 로그 프레임워크에서는 FATAL 이라는 최고 수준의 로그 레벨을 제공하기도 하지만, 로그백에서 FATAL은 ERROR로 매핑되니, FATAL이라는 수준이 있구나 정도만 아셔도 될 것 같습니다.)~~

### 로컬 환경
콘솔에 출력

spring boot -> info
application logic -> debug
hibernate -> debug

### 개발 환경
파일로 출력

전체 -> info

### 운영 환경 (현재 분리되지 않은 환경으로, 논의 필요)
파일로 출력

전체 -> error

**Comment by zeus6768** [comment]
### 안건

운영 환경에서 처리된 예외(CodeZapException)를 error 미만 등급으로 로깅하면 좋겠어요.

### 근거

1. 운영환경에서 어떤 예외가 발생하는지 수집해서, 사용자의 활동을 분석하는 용도로 사용할 수 있습니다. 

2. 프론트엔드 팀에서 운영 서버의 로그를 요청하는 경우가 있었습니다.
